### PR TITLE
Add reader_method to GlobReaderGenerator (#914)

### PR DESCRIPTION
**Pull request by cwerner**
This is my attempt to add the optional reader_method argument from SubdirReaderGenerator to GlobReaderGenerator.

This should close issue #914.

### CODE REVIEW

**Comment by kyleaton** [comment]
@clabot check

**Comment by kyleaton** [comment]
@cla-bot check

**Review by jcampbell — CHANGES_REQUESTED** [review]
This looks great, really excited to merge it! Three very small requests:

1. Could you also allow the reader_method to be configured on a per-asset-glob basis, with the logic being that the base generator definition is overridden with a per-asset configuration if one is present?

2. Would you add a reader_method key to the example configuration in the docstring?

3. Would you add a reader_method to the tested config in the test_glob_reader_generator method of the test_batch_generators file?

**Comment by cwerner** [comment]
Hi

Regarding your comments:
1: I tried that but I got lost in the hierarchy of function calls. I'm not sure where the args are actually extracted? I tried to trace it down but failed.
2: Done (maybe the per-asset line is confusing and should be explained (reads like a duplicate to the global setting, but I did not want to introduce explanatory text either...)
3: here, I'd also need some more pointers/ help. not sure how/ where to do this.

File: great_expectations/datasource/generator/glob_reader_generator.py [hunk 1]
```diff
@@ -119,7 +127,7 @@ def build_batch_kwargs_from_partition_id(self, generator_asset, partition_id=Non
                                         partition_id: partition_id
                                     })
         batch_kwargs = self._build_batch_kwargs_from_path(path[0], glob_config, reader_options=reader_options,
-                                                          limit=limit, partition_id=partition_id)
+                                                          limit=limit, reader_method=reader_method, partition_id=partition_id)
```
Comment (jcampbell) [inline great_expectations/datasource/generator/glob_reader_generator.py]:
```suggestion
                                                          limit=limit, partition_id=partition_id)
```

Comment (jcampbell) [reply great_expectations/datasource/generator/glob_reader_generator.py]: I think between `self.reader_method` and `glob_config`, you've already got the relevant info here.

File: great_expectations/datasource/generator/glob_reader_generator.py [hunk 2]
```diff
@@ -156,11 +164,11 @@ def _get_iterator(self, generator_asset, reader_options=None, limit=None):
         paths = glob.glob(os.path.join(self.base_directory, glob_config["glob"]))
         return self._build_batch_kwargs_path_iter(paths, glob_config, reader_options=reader_options, limit=limit)
 
-    def _build_batch_kwargs_path_iter(self, path_list, glob_config, reader_options=None, limit=None):
+    def _build_batch_kwargs_path_iter(self, path_list, glob_config, reader_options=None, limit=None, reader_method=None):
```
Comment (jcampbell) [inline great_expectations/datasource/generator/glob_reader_generator.py]:
```suggestion
    def _build_batch_kwargs_path_iter(self, path_list, glob_config, reader_options=None, limit=None):
```

I think between `self.reader_method` and `glob_config`, you've already got the relevant info here.

File: great_expectations/datasource/generator/glob_reader_generator.py [hunk 3]
```diff
@@ -156,11 +164,11 @@ def _get_iterator(self, generator_asset, reader_options=None, limit=None):
...
         for path in path_list:
-            yield self._build_batch_kwargs_from_path(path, glob_config, reader_options=reader_options, limit=limit)
+            yield self._build_batch_kwargs_from_path(path, glob_config, reader_options=reader_options, limit=limit, reader_method=None)
```
Comment (jcampbell) [inline great_expectations/datasource/generator/glob_reader_generator.py]: I think between `self.reader_method` and `glob_config`, you've already got the relevant info here.
```suggestion
            yield self._build_batch_kwargs_from_path(path, glob_config, reader_options=reader_options, limit=limit)
```

File: great_expectations/datasource/generator/glob_reader_generator.py [hunk 4]
```diff
@@ -156,11 +164,11 @@ def _get_iterator(self, generator_asset, reader_options=None, limit=None):
...
 
-    def _build_batch_kwargs_from_path(self, path, glob_config, reader_options=None, limit=None, partition_id=None):
+    def _build_batch_kwargs_from_path(self, path, glob_config, reader_options=None, limit=None, reader_method=None, partition_id=None):
```
Comment (jcampbell) [inline great_expectations/datasource/generator/glob_reader_generator.py]:
```suggestion
    def _build_batch_kwargs_from_path(self, path, glob_config, reader_options=None, limit=None, partition_id=None):
```
ditto above

File: great_expectations/datasource/generator/glob_reader_generator.py [hunk 5]
```diff
@@ -189,6 +197,12 @@ def _build_batch_kwargs_from_path(self, path, glob_config, reader_options=None,
         if limit is not None:
             batch_kwargs['limit'] = limit
 
+        if self.reader_method is not None:
+            batch_kwargs['reader_method'] = self.reader_method
+        
+        if reader_method:
+            batch_kwargs['reader_method'].update(reader_method)
```
Comment (jcampbell) [inline great_expectations/datasource/generator/glob_reader_generator.py]:
```suggestion
        if glob_config.get("reader_method"):
                batch_kwargs['reader_method'] = glob_config.get("reader_method")
```

Comment (jcampbell) [reply great_expectations/datasource/generator/glob_reader_generator.py]: Not sure a multiline suggestion worked there...

**Review by jcampbell — CHANGES_REQUESTED** [review]
This looks great, really excited to merge it! Three very small requests:

1. Could you also allow the reader_method to be configured on a per-asset-glob basis, with the logic being that the base generator definition is overridden with a per-asset configuration if one is present?

2. Would you add a reader_method key to the example configuration in the docstring?

3. Would you add a reader_method to the tested config in the test_glob_reader_generator method of the test_batch_generators file?

**Comment by jcampbell** [comment]
I think my comment may have made it seem more complicated, but I added a snippet in the comment that I think will use the glob_config to override the the global config if present.

**Comment by jcampbell** [comment]
For the third comment:

In the file `tests/datasource/generator/test_glob_reader_generator.py` there's a method called `test_glob_reader_generator_partitioning`...there is both a GlobReaderGenerator configuration and several asset configurations there; could you add reader_method to one of the asset configurations and also to the top-level configuration, and then add an assertion about the correctness of the resulting batch_kwargs?

**Review by jcampbell — APPROVED** [review]

**Pull request merged**